azure/services/identities/client.go

```
@@ -0,0 +1,82 @@
+}
+
+// azureClient contains the Azure go-sdk Client.
+type azureClient struct {
+	userAssignedIdentities msi.UserAssignedIdentitiesClient
+}
+
+var _ client = (*azureClient)(nil)
+
+// newClient creates a new MSI client from subscription ID.
+func newClient(auth azure.Authorizer) *azureClient {
+	c := newUserAssignedIdentitiesClient(auth.SubscriptionID(), auth.BaseURI(), auth.Authorizer())
+	return &azureClient{c}
+}
+
+// newIdentitiesClient creates a new user-assigned identities client from subscription ID.
+func newUserAssignedIdentitiesClient(subscriptionID string, baseURI string, authorizer autorest.Authorizer) msi.UserAssignedIdentitiesClient {
+	userAssignedIdentitiesClient := msi.NewUserAssignedIdentitiesClientWithBaseURI(baseURI, subscriptionID)
+	azure.SetAutoRestClientDefaults(&userAssignedIdentitiesClient.Client, authorizer)
+	return userAssignedIdentitiesClient
+}
+
+// Get the managed service identity.
+func (ac *azureClient) Get(ctx context.Context, resourceGroupName, name string) (msi.Identity, error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "identities.azureClient.Get")
+	defer done()
+
+	return ac.userAssignedIdentities.Get(ctx, resourceGroupName, name)
+}
+
+// GetClientID returns the client ID of a managed service identity, given its full URL identifier.
+func GetClientID(ctx context.Context, auth azure.Authorizer, providerID string) string {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "identities.GetClientID")
+	defer done()
+
+	// parse providerID into resource group and MSI name
+	rgAndName := regexp.MustCompile(`^/subscriptions/[\w-]+/resourceGroups/(?P<rg>.+)/providers/Microsoft.ManagedIdentity/userAssignedIdentities/(?P<name>.+)$`)
+	matches := rgAndName.FindStringSubmatch(providerID)
+	if len(matches) < 3 {
+		// TODO: log error and return an empty string
+		return "failed to find resource group and name in provider ID URL"
```

This should return an error, we can handle the error from the controller